src/test/java/com/mongodb/hibernate/jdbc/MongoConnectionProviderTests.java

```diff
@@ -102,6 +103,49 @@ void testMongoClientCustomizerThrowException() {
                         null)) {}
             });
         }
+
+        private static class TestClusterListener implements ClusterListener {
+            private static final Duration TIMEOUT = Duration.ofSeconds(5);
+
+            private final CompletableFuture<ClusterOpeningEvent> openingEvent;
+            private final CompletableFuture<ClusterClosedEvent> closedEvent;
+
+            TestClusterListener() {
+                openingEvent = new CompletableFuture<>();
+                closedEvent = new CompletableFuture<>();
+            }
+
+            @Override
+            public void clusterOpening(ClusterOpeningEvent event) {
+                assertTrue(openingEvent.complete(event));
+            }
+
+            @Override
+            public void clusterClosed(ClusterClosedEvent event) {
+                assertTrue(closedEvent.complete(event));
+            }
+
+            void assertClusterOpening() {
+                assertDone(openingEvent);
+            }
+
+            void assertClusterClosed() {
+                assertDone(closedEvent);
+            }
+
+            private static void assertDone(final Future<?> future) {
```

Comment:
`final` is not necessary in spotless and errorprone context, but it is fine to use it in this project.

Reply:
It's our style in the driver code, and I failed to "switch" the style I am supposed to use. Sorry, fixed in 430e9ced91d295ba3b0b46d70dbf7d04b0c948d6.

Reply:
Style issue is not that vital to my taste (compared to the complex Hibernate integration business code logic) and it is not a bad idea to allow for some flexibility and personal preference.